### ISSUE TITLE
Inconsistent Scheduling Using Block Intervals

### ISSUE THREAD

**Issue by medariox** [ponder-sh/ponder#1122]
The current approach of using [block intervals](https://ponder.sh/docs/block-intervals) to simulate cron-like scheduling assumes that blocks are consistently produced (cit. "run at a consistent frequency") at regular intervals (e.g., every 12 seconds on Ethereum). However, as shown in the chart below from Etherscan, **the average block production time is not exactly 12s**. According to the [Ethereum documentation](https://ethereum.org/en/developers/docs/blocks/#block-time), block times are only approximately 12s. In fact, the current average block time is 12.05s, as reported by [YCharts](https://ycharts.com/indicators/ethereum_average_block_time).

Furthermore, this issue is not exclusive to Ethereum. Other blockchains also face similar inconsistencies in block production times, as shown by data from [this Dune dashboard](https://dune.com/jacobdcastro/avg-block-times). This variability causes issues when scheduling tasks using block intervals.

This issue is especially noticeable when indexing longer periods, where tasks are expected to run at consistent timestamps. The current implementation leads to unreliable schedules and drift, which makes it unsuitable for use cases requiring strict time-based triggers.

These are the main problems with the current implementation:

- **Inconsistent Trigger Times**: Tasks scheduled to run at regular block intervals do not correspond to consistent real-world times. For example, a task intended to run every 12s may trigger at varying intervals depending on actual block production times. The chart shows how block times deviate from the expected 12s, causing tasks to run at irregular intervals.

 - **Schedule Drift Over Time**: Over longer periods, the cumulative inconsistency in block production causes task execution times to "drift," making it unreliable for tasks that require precise timing (e.g., running every day at 12:00 PM).

- **Misalignment with Real-World Time**: For use cases that require strict adherence to real-world timeframes (e.g., every minute, hour, or day), block intervals do not guarantee accurate synchronization, making this approach unsuitable for time-sensitive applications.

**Proposed Solution**
Consider an alternative mechanism to track real-world time or provide a way to trigger tasks based on wall-clock time, independent of block production intervals. One possible solution is to **use the block timestamp** instead of relying solely on block intervals. By referencing the block's timestamp, tasks could be scheduled based on real-world time, reducing the drift and inconsistency introduced by irregular block production.

This would allow for more consistent and reliable cron-like scheduling, especially for applications that depend on precise timing.

Thank you to the Ponder maintainers for all your hard work and for taking the time to review this issue. 🙏

**Comment by 0xOlias** [comment]
Thanks for the thoughtful writeup!

Definitely acknowledge this limitation, and it would be great to find a suitable fix. For context, we considered using a more traditional cron feature based on timestamps, but decided on the current block based approach because it was much simpler internally, and seemed to solve 90%+ of the use cases we heard from users around this. Another consideration is that OP stack chains (which I'd estimate account for ~50% of Ponder usage rn) do not have a variable timestamp, so they don't suffer from the problems above.

If you can share a few specific uses cases where block intervals do not work but exact timestamp/cron would, that would help us make progress on this.

**Comment by medariox** [comment]
Thank you for the quick response!

My primary use case is to chart the last hourly price of a token. Based on the documentation, I believe my current option is to configure the scheduler like this:
```
blocks: {
    HourlyScheduler: {
      network: "mainnet",
      startBlock: 20421017, // Timestamp: 1722362399 (Jul 30 2024 17:59:59 GMT+0000)
      interval: (60 * 60) / 12, // Every 60 minutes
    },
  },
```

However, due to the issues mentioned earlier, this approach isn't ideal. For reference, here’s an example of block timestamps:
```
.....
Block: 20803585 | Timestamp: 1726977599 // Jul 22 2024 03:59:59 GMT+0000
Block: 20803885 | Timestamp: 1726981223 // Jul 22 2024 05:00:23 GMT+0000
Block: 20804185 | Timestamp: 1726984835 // Jul 22 2024 06:00:35 GMT+0000
.....
```

As you can see, the blocks don’t align perfectly with the hourly schedule. But as a matter of fact, there are blocks produced at these exact timestamps:
```
.....
Block: 20803585 | Timestamp: 1726977599 // Jul 22 2024 03:59:59 GMT+0000
Block: 20803883 | Timestamp: 1726981199 // Jul 22 2024 04:59:59 GMT+0000
Block: 20804182 | Timestamp: 1726984799 // Jul 22 2024 05:59:59 GMT+0000
.....
```

This could be achieved if we could configure it like this:
```
blocks: {
    HourlyScheduler: {
      network: "mainnet",
      startBlock: 20421017, // Timestamp: 1722362399 (Jul 30 2024 17:59:59 GMT+0000)
      timestampInterval: 3600, // Every 60 minutes
    },
  },
```

I’m not entirely sure which use cases the current block interval option is meant to solve, but it seems clear that it’s not well-suited for scheduling tasks at consistent real-time frequencies, even though the documentation suggests it e.g. with the example: `(60 * 60) / 12 // Every 60 minutes`.

I hope this clarifies my use case and why I think it’s important. Please let me know if there’s anything I can do to help speed this along. Thanks again!

**Comment by frontier159** [comment]
@typedarray would you know if this is scheduled by anyone at the moment? Agree it would be a super useful feature that (AFAIK) isn't available in other solutions

**Comment by typedarray** [comment]
The main blocker to shipping this is a design decision.

Quick background: All indexing functions today have a "checkpoint" value, which consists of a timestamp, chain ID, block number, and an "execution index" within the block (eg transaction index, log index). The checkpoint serves two critical functions, 1) event ordering and 2) block number used in `context.client` RPC requests.

For the proposed "cron" events, there's a clear path to solving the event ordering issue. We can just use the timestamp generated by the cron, and set the other fields to zero - so within a given second, a cron event would always run before any other events that happen to have a block timestamp at that same second (at the top of the second) or after them (bottom of the second).

However, we can't figure out what to do about the block number to use in `context.client` RPC requests. Should it be whichever block number on the target chain is <= the cron timestamp? <= ? Disallow `context.client` in cron indexing functions (yuck)?

And even if we can decide on a design, the implementation is not trivial - Ponder doesn't constantly maintain a mapping of block timestamps -> block numbers across all chains. But, we could probably get clever and figure something out that doesn't add too many RPC requests.

Appreciate user perspectives on how you'd expect this to work. BTW, this is the same core problem blocking https://github.com/ponder-sh/ponder/issues/1446.

**Comment by frontier159** [comment]
@typedarray sorry I missed this reply.

I think there is value in using real block metadata here - what if the cron scheduler isn't based on wall clock time of the system running ponder, but based on actual block timestamps?

> The "cron" event will fire when a block is observed that has a timestamp greater than or equal to the cron schedule time.
> The event metadata (aka `checkpoint`?) will represent that observed block >= the cron schedule time.

So as an example:
* Given a cron schedule: eg `15 0 * * * *` (the fifteenth second of every hour)
* For ETH mainnet which has (roughly) 12 second block times, assuming that falls on `00:00:00`, `00:00:12`, `00:00:24`, ...

That means the event wont fire until after the block mined at `00:00:24` is observed. It will have checkpoint info of:
```
{
  timestamp: 00:00:24
  chainId: 0 (or perhaps nominated on the cron schedule config if useful for clients)
  blockNumber: The real observed block number at 00:00:24
  executionIndex: -1, // Always schedule this first before actual events within the block
}
```

wrt `executionIndex`, I think it should either be at the start or at the end of the block. imo whatever is easiest for implementation.

One goal would be that if we reindex for past blocks (resync on the same codebase) we end up with the same records in the database. I believe the above would achieve this, because the events fire off based on the actual block times.

Would this design also help with the concerns on maintaining the block timestamps => block numbers? I hope it would avoid having to add any extra mappings at all, but I'm not familiar with the codebase.

**Comment by frontier159** [comment]
gm @typedarray just checking to see if this makes sense ☝️

**Comment by julien51** [comment]
It's not quite exactly the issue reported but I was wondering if there was a way to have an `interval` be chain specific. Our application indexes multiple chains, but each chain has a very different block time (Ethereum is 12s and Base is 0.2s). Having a "single" `interval` means that by the time we get 1 handler called for Ethereum, we have 60 Base handler calls. Maybe `interval` could be added "inside" the network object, along with `startBlock`?

**Comment by brunowowk** [comment]
Probably relevant:
[EIP-7782](https://eips.ethereum.org/EIPS/eip-7782) will likely be included in Glamsterdam, so mainnet block times will decrease from ~12s to ~6s, which will make block interval scheduling pretty unreliable (different sampling rates before and after the change)

It would be great to have proper cron scheduling before that change drops; otherwise, indexing data at regular intervals on mainnet will get pretty tricky.

**Comment by typedarray** [comment]
> Maybe interval could be added "inside" the network object, along with startBlock?

@julien51 this is already supported!

```ts
import { createConfig } from "ponder";

export default createConfig({
  chains: {
    mainnet: {
      id: 1,
      rpc: process.env.PONDER_RPC_URL_1,
    },
    base: {
      id: 8453,
      rpc: process.env.PONDER_RPC_URL_8453,
    },
  },
  blocks: {
    ChainlinkPriceOracle: {
      chain: {
        mainnet: {
          startBlock: 19_750_000,
          interval: 60 / 12, // every minute
        },
        base: {
          startBlock: 26_000_000,
          interval: 60 / 2, // every minute
        },
      },
    },
  },
});
```

**Comment by frontier159** [comment]
@typedarray did you have feedback on this proposal? https://github.com/ponder-sh/ponder/issues/1122#issuecomment-3034423239

**Comment by kyscott18** [comment]
This feature is out-of-scope for ponder
